amazon/ion/ioncmodule.c

```diff
@@ -994,31 +1038,50 @@ static iERR ionc_read_timestamp(hREADER hreader, PyObject** timestamp_out) {
         case ION_TS_FRAC:
         {
             decQuad fraction = timestamp_value.fraction;
+            decQuad tmp;
+
             int32_t fractional_precision = decQuadGetExponent(&fraction);
             if (fractional_precision > 0) {
                 _FAILWITHMSG(IERR_INVALID_TIMESTAMP, "Timestamp fractional precision cannot be a positive number.");
             }
             fractional_precision = fractional_precision * -1;
 
-            decQuad tmp;
-            decQuadScaleB(&fraction, &fraction, decQuadFromInt32(&tmp, MICROSECOND_DIGITS), &dec_context);
-            int32_t microsecond = decQuadToInt32Exact(&fraction, &dec_context, DEC_ROUND_DOWN);
             if (fractional_precision > MICROSECOND_DIGITS) {
-                // C extension only supports up to microsecond precision.
-                fractional_precision = MICROSECOND_DIGITS;
-            }
+                decQuadScaleB(&fraction, &fraction, decQuadFromInt32(&tmp, fractional_precision), &dec_context);
+                int dec = decQuadToInt32Exact(&fraction, &dec_context, DEC_ROUND_DOWN);
+                if (fractional_precision > MAX_TIMESTAMP_PRECISION) fractional_precision = MAX_TIMESTAMP_PRECISION;
+                if (decContextTestStatus(&dec_context, DEC_Inexact)) {
+                    // This means the fractional component is not [0, 1) or has more than microsecond precision.
+                    decContextClearStatus(&dec_context, DEC_Inexact);
+                }
 
-            if (decContextTestStatus(&dec_context, DEC_Inexact)) {
-                // This means the fractional component is not [0, 1) or has more than microsecond precision.
-                decContextClearStatus(&dec_context, DEC_Inexact);
-            }
-            PyObject* py_fractional_precision = PyInt_FromLong(fractional_precision);
-            PyObject* py_microsecond = PyInt_FromLong(microsecond);
-            PyDict_SetItemString(timestamp_args, "fractional_precision", py_fractional_precision);
-            PyDict_SetItemString(timestamp_args, "microsecond", py_microsecond);
-            Py_DECREF(py_fractional_precision);
-            Py_DECREF(py_microsecond);
+                char* front = "0.";
+                char decimal[fractional_precision];
+                sprintf(decimal, "%d", dec);
+                char* dec_num = string_concatenation(2, fractional_precision, front, decimal);
+
+                PyObject* py_dec_str = PyUnicode_FromString(dec_num);
+                PyObject* py_fractional_seconds =
+                    PyObject_CallFunctionObjArgs(_decimal_constructor, py_dec_str, NULL);
+                PyDict_SetItemString(timestamp_args, "fractional_seconds", py_fractional_seconds);
+                Py_DECREF(py_fractional_seconds);
+                Py_DECREF(py_dec_str);
```

Comment:
Rather than creating a temporary string to be re-parsed by the decimal constructor, we should just create a decimal from the `dec` integer, then `scaleb` it using `fractional_precision`.

Reply:
Changed. Passed unit test.